chore(docs): New "Creating a Source Plugin" tutorial (#37538)
gatsbyjs/gatsby: docs/docs/tutorial/creating-a-source-plugin/part-1/index.mdx

<Collapsible
  summary={<em>GitHub cloning instructions</em>}
>

HTTPS snippet:

```shell
git clone https://github.com/gatsbyjs/creating-source-plugin-tutorial.git
```

SSH snippet:

```shell
git clone git@github.com:gatsbyjs/creating-source-plugin-tutorial.git
```

Also be sure to check out [GitHub's documentation for how to clone a repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository).

</Collapsible>

After successfully cloning the repository, navigate to its folder and switch to the `start` branch to have a clean boilerplate:

```shell
git checkout start
```

Open the repository in your favorite editor as it'll make the next step easier.

## Project structure

The `creating-source-plugin-tutorial` project includes three directories:

- `api`: This is the example backend API
- `plugin`: The source plugin
- `site`: An example Gatsby site

> So the `site` uses the `plugin` which sources its data from `api`. As mentioned in [Part 0](/docs/tutorial/creating-a-source-plugin/part-0/#development-environment) this project uses yarn workspaces to make this setup work.

Reviewer: ```suggestion
The `site` uses the `plugin` which sources its data from `api`. As mentioned in [Part 0](/docs/tutorial/creating-a-source-plugin/part-0/#development-environment) this project uses yarn workspaces to make this setup work.
```

By having the plugin and an example site that uses the plugin in the same repository, you can more easily test your plugin.

<Announcement>

You can ignore the `api` folder unless you're curious how a [GraphQL Yoga server](https://github.com/dotansimha/graphql-yoga) is set up. For the purpose of this tutorial the example backend API was included so that this tutorial doesn't rely on outside resources. This way you can run everything locally.

</Announcement>

The `plugin` itself has a structure we'd also recommend following:

- A `package.json` file that sets `gatsby` and `gatsby-plugin` as its `"keywords"` so that Gatsby's plugin search can pick it up

> - A root `gatsby-node.js` that requires `./dist/gatsby-node` (the compiled TypeScript file). At the moment Gatsby looks for a root `gatsby-node.js` file as its entry file.

Reviewer 2: I noticed some of the list items here end with a '.' and some don't. Is it if there are more than one sentence, then a period is required? Just wondering what approach the Gatsby docs use!

> - A well-written `README` that explains the user in concise steps how to install, use, and configure your plugin (also see [How to write a plugin README](/contributing/docs-contributions/how-to-write-a-plugin-readme/))

Reviewer 2: Possible typo:
```suggestion
- A well-written `README` that explains to the user in concise steps how to install, use, and configure your plugin (also see [How to write a plugin README](/contributing/docs-contributions/how-to-write-a-plugin-readme/))
```

- A `src` folder with the plugin's source code
  - A root `gatsby-node.ts` file that exports [Gatsby Node APIs](/docs/reference/config-files/gatsby-node/) from individual files
  - Individual files that only contain the specific Node API (in [kebab case](https://en.wikipedia.org/wiki/Letter_case#Kebab_case)), e.g. `onPluginInit` becomes `on-plugin-init.ts`. This will also make unit testing your plugin easier.

> How you organize your `src` folder is of course up to you, we'd recommend keeping things consistent and files lowercase.

Reviewer 2: ```suggestion
How you organize your `src` folder is of course up to you, but we'd recommend keeping things consistent and files lowercase.
```

<Collapsible
  summary={<em>Wondering about a source plugin template for new projects?</em>}
>

> Above we said that you can ignore the `api` folder. This is because for your real-world project you'll most probably source the data from an external API, not an API locally in your repository.

Reviewer: ```suggestion
Above we said that you can ignore the `api` folder. This is because for your real-world project you'll most likely source the data from an external API, not an API locally in your repository.
```

Reviewer 2: +1

The companion repository not only has a `start` branch but also a `template` branch without the `api` folder. You'll learn more about this at the end in the [What's next?](/docs/tutorial/creating-a-source-plugin/whats-next/) part. For now, go through the tutorial and at the end you'll know how to leverage the repository as a template for your new projects!

</Collapsible>

## Start the project

Ready to see something happen in your browser? Then let's go!

First, install the necessary dependencies to run everything:

```shell
yarn
```

Afterwards, open a new window in your terminal (you should have two now).

> In the first window, run `develop:deps` to start watching the `api` and `plugin` directories. If you change something inside `api`, the GraphQL server will be restarted, if you change something in `plugin` the TypeScript compiler will output updated files.

Reviewer: ```suggestion
In the first window, run the `develop:deps` script to start watching the `api` and `plugin` directories. If you change something inside `api`, the GraphQL server will be restarted, if you change something in `plugin` the TypeScript compiler will output updated files.
```

```shell
yarn develop:deps
```

> You should see something like this in your terminal:

Reviewer 2: Suggestion to update to swap out a sighted term ("see"):
```suggestion
You should find something like this in your terminal:
```

```shell
[develop:api   ] 14:48:03 - Starting compilation in watch mode...
[develop:api   ]
[develop:plugin] 14:48:03 - Starting compilation in watch mode...
[develop:plugin]
[develop:api   ] Server is running at http://localhost:4000/graphql
[develop:api   ]
[develop:api   ] 14:48:04 - Found 0 errors. Watching for file changes.
[develop:plugin]
[develop:plugin] 14:48:04 - Found 0 errors. Watching for file changes.
[develop:api   ] Server is running at http://localhost:4000/graphql
```

> Now, in your second terminal window, run `develop:site`. This runs `gatsby develop` for the example site:

Reviewer: ```suggestion
Now, in your second terminal window, run the `develop:site` script. This runs `gatsby develop` for the example site:
```

Reviewer: (with these, I think it's best to make it clear that a user should not just type `develop:site` into their terminal, and that it is a package.json script

```shell
yarn develop:site
```

> While `gatsby develop` is running you should also see the `info Example plugin loaded...` log. This shows that the plugin was successfully loaded for this initial demo. Yey 🎉

Reviewer 2: Small typo and a suggestion you can take or leave to avoid using sighted terms:
```suggestion
While `gatsby develop` is running you should also note the `info Example plugin loaded...` log. This verifies that the plugin was successfully loaded for this initial demo. Yay 🎉
```

</Announcement>

Once the development server is ready, go to `http://localhost:8000`, and be greeted by a minimalistic index page.

![A screenshot of "localhost:8000" in a web browser. There's a heading that says, "All posts" and a paragraph that says, "Posts will go here"](./01-clean-index-page.png)

### Key takeaways

- It's good practice to develop your source plugin alongside an example site to be able to test it
- You can keep your code organized by placing each Gatsby Node API in its own file
- When making changes to your plugin, you'll need to restart `gatsby develop` to see your changes applied

<Announcement>

**Share Your Feedback!**

Our goal is for this tutorial to be helpful and easy to follow. We'd love to hear your feedback about what you liked or didn't like about this part of the tutorial.

Use the "Was this doc helpful to you?" form at the bottom of this page to let us know what worked well and what we can improve.

</Announcement>

### What's coming next?

> In Part 2 you'll learn how to do the most important part of a Gatsby source plugin: Sourcing data and displaying it on your site.

Reviewer 2: Suggestion to change:
```suggestion
In Part 2 you'll learn how to do the most important part of a Gatsby source plugin: sourcing data and displaying it on your site.
```
